rfcs/0000-multi-server-course-content.md

```diff
@@ -0,0 +1,58 @@
+
+When course content is synced, it will be broken up into a number of chunks:
+
+* One chunk containing all course custom elements
+* One chunk for `clientFilesCourse`
+* One chunk for `serverFilesCourse`
+* A chunk for each course instance's `clientFilesCourseInstance`
+* A chunk for each course instance's `serverFilesCourseInstance`
+* One chunk for each question
+* A chunk for all question thumbnails
+
+These chunks will be uploaded to some file store with a unique, possibly deterministic ID.
+
+When a server needs to use some course content, it will download the necessary chunks to disk. Here are some common pages and the chunks necessary for them:
+
+* Viewing the course question page
+  * Question thumbnails chunk
+* Viewing or grading a question
+  * Custom elements chunk
+  * Question chunk
+  * `serverFilesCourse` chunk
+  * Appropriate `serverFilesCourseInstance` chunk
+* Serving a file from `clientFilesCourse`
+  * `clientFilesCourse` chunk
+
+Note that in many cases, the chunk will have already been loaded to disk for a previous request, and so it can be reused. In fact, this is what makes this method so attractive: it allows us to amortize the cost of loading a course's content over the course of many requests. Files are loaded on-demand and are aggressively cached locally, meaning that requests will only be slightly slower (on the order of hundreds of milliseconds) when they requires course chunks that don't yet exist locally on disk.
+
+## Syncing
+
+In order to sync course content, we'll still need entire course repos to exist *somewhere* on disk. To avoid the overhead of needing to clone an entire course repo for each sync, we'll dedicate a single PrairieLearn instance solely to syncing operations. This instance will be like the current PrairieLearn server in that it will contain all course repos on disk. We'll configure a load balancer to route all syncing requests to that instance. That instance will then sync metadata to the database and upload the course's chunks as appropriate. Note that this means we don't need a separate service to handle syncing - when running locally, all syncing requests will go to the same server as usual. And servers running in production don't even need to be aware of whether or not they're the syncing instance. This keeps the overall design simple.
+
+When a course instance is synced, we update its git hash in the database. This hash will serve as a unique version identifier for the course's contents. The next time someone tries to access something from the course, we'll check if we already have a working directory for that particular version. If not, we can create one that will store only files for the particular revision, like `/courses/my_course_id/234fae...[hash]...9823efde/`. We'll then start loading files again to this new directory so as not to disrupt any question renders that might be in progress. We can then have a background task that periodically deletes old versions.
```

Comment:
One thing I don't quite understand is the distinction between the course-level hash, which uniquely defines the whole course, and chunk-level hashes, which uniquely define the chunk. Most of the time, updating a course (and getting a new hash for it) will not change most of the chunk-level hashes, so it would be a shame to have to re-download them all.

I'm assuming that we store a DB table that records which version (hash) of each chunk we have on each machine, as well as the hash of the latest version of each chunk that's available in S3? This way each machine can run a fast query to make sure that it has up-to-date versions of all needed chunks before it starts processing the query.

When making local copies of updated chunks, it seems to me we could either:
1. Unpack them in a temporary directory and then atomically move them into place in a fixed course directory (without the hash), like `/courses/my_course_id/`.
2. Use course directories with the course-hash in the name (`/courses/my_course_id/234fae.../`) but when we make a new one of these, we hard-link all existing content from the last version so we don't have to re-download existing chunks.

While option (2) is more robust, I'm actually not sure it's needed, especially in a first version. It would be easy to retrofit later. We've been running with in-place updates from sync (option 1) for many years without any noticeable problems.

Reply:
I agree that in-place updates should be sufficient, especially since we've been doing it that way since the start. Long-term, I think we could actually do truly atomic updates by introducing some new APIs. Currently, questions must use hardcoded paths like `../../serverFilesCourse/foo/bar.txt` to access assets. If we instead provided an API to questions like `data['options']['server_files_course_path']` to complement existing things like `data['options']['client_files_question_url']`, we could unpack new server files to `/courses/my_course_id/serverFilesCourse-1234fdea...[hash]...4321` instead of just `/courses/my_course_id/serverFilesCourse`. This way, we could keep the files seen by a particular question render/grade/etc. wholly consistent! This would be very backwards incompatible though, so this is definitely a future optimization we could potentially make.

Reply:
I was thinking about this further and I’m actually a bit tempted to “do it right” from the start with respect to atomic update, if it isn’t too complex. You mentioned earlier about using symlinks to activate new components. This seems relatively easy and robust?